src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs

```diff
@@ -175,50 +175,83 @@ private BoundAttribute BindAttributeCore(AttributeSyntax node, NamedTypeSymbol a
-                                                                out boundConstructorArguments);
-            }
-            else
+            if (attributeTypeForBinding.IsErrorType())
             {
                 boundConstructorArguments = analyzedArguments.ConstructorArguments.Arguments.SelectAsArray(
                     static (arg, attributeArgumentBinder) => attributeArgumentBinder.BindToTypeForErrorRecovery(arg),
                     attributeArgumentBinder);
             }
-            Debug.Assert(boundConstructorArguments.All(a => !a.NeedsToBeConverted()));
-            diagnostics.Add(node, useSiteInfo);
-
-            if (attributeConstructor is object)
+            else
             {
-                ReportDiagnosticsIfObsolete(diagnostics, attributeConstructor, node, hasBaseReceiver: false);
-
-                if (attributeConstructor.Parameters.Any(p => p.RefKind == RefKind.In))
+                bool found = TryPerformConstructorOverloadResolution(
+                    attributeTypeForBinding,
+                    analyzedArguments.ConstructorArguments,
+                    attributeTypeForBinding.Name,
+                    node.Location,
+                    attributeType.IsErrorType(),
+                    diagnostics,
+                    out var memberResolutionResult,
+                    out var candidateConstructors,
+                    allowProtectedConstructorsOfBaseType: true);
+                attributeConstructor = memberResolutionResult.Member;
+                expanded = memberResolutionResult.Resolution == MemberResolutionKind.ApplicableInExpandedForm;
+                argsToParamsOpt = memberResolutionResult.Result.ArgsToParamsOpt;
+
+                if (!found)
                 {
-                    Error(diagnostics, ErrorCode.ERR_AttributeCtorInParameter, node, attributeConstructor.ToDisplayString(SymbolDisplayFormat.CSharpErrorMessageFormat));
+                    resultKind = resultKind.WorseResultKind(
+                        memberResolutionResult.IsValid && !IsConstructorAccessible(memberResolutionResult.Member, ref useSiteInfo) ?
+                            LookupResultKind.Inaccessible :
+                            LookupResultKind.OverloadResolutionFailure);
+                    boundConstructorArguments = BuildArgumentsForErrorRecovery(analyzedArguments.ConstructorArguments, candidateConstructors);
+                }
+                else
+                {
+                    attributeArgumentBinder.BindDefaultArguments(
+                        node,
+                        attributeConstructor.Parameters,
+                        analyzedArguments.ConstructorArguments.Arguments,
+                        argumentRefKindsBuilder: null,
+                        ref argsToParamsOpt,
+                        out defaultArguments,
+                        expanded,
+                        enableCallerInfo: !IsEarlyAttributeBinder,
```

Comment:
Anything will break if we always passed `true`?

Reply:
This was to try and minimize behavior changes when wiring everything up. I wasn't able to identify a case where behavior changed when we passed `true` here. I think you would need to have certain specific well-known attribute declarations with caller info attributes on optional parameters, then use the attribute with implicit default arguments. Not something to spend a ton of effort defending against, IMO.